src/alchemlyb/workflows/abfe.py

```diff
@@ -0,0 +1,665 @@
+        self.logger.info(f'Set Software to {software}.')
+
+        self.update_units(units)
+
+        self.logger.info(f'Finding files with prefix: {prefix}, suffix: '
+                         f'{suffix} under directory {dir} produced by '
+                         f'{software}')
+        self.file_list = glob(dir + '/**/' + prefix + '*' + suffix,
+                         recursive=True)
+
+        self.logger.info(f'Found {len(self.file_list)} xvg files.')
+        self.logger.info("Unsorted file list: \n %s", '\n'.join(
+            self.file_list))
+
+        if software.lower() == 'gromacs':
+            self.logger.info(f'Using {software} parser to read the data.')
+            self._extract_u_nk = gmx.extract_u_nk
+            self._extract_dHdl = gmx.extract_dHdl
+        elif software.lower() == 'amber':
+            self._extract_u_nk = amber.extract_u_nk
+            self._extract_dHdl = amber.extract_dHdl
+        else: # pragma: no cover
+            raise NameError(f'{software} parser not found.')
+
+    def read(self):
+        '''Read the u_nk and dHdL data from the
+        :attr:`~alchemlyb.workflows.ABFE.file_list`
+
+        Attributes
+        ----------
+        u_nk_list : list
+            A list of :class:`pandas.DataFrame` of u_nk.
+        dHdl_list : list
+            A list of :class:`pandas.DataFrame` of dHdl.
+        '''
+        u_nk_list = []
+        dHdl_list = []
+        for xvg in self.file_list:
+            try:
+                u_nk = self._extract_u_nk(xvg, T=self.T)
+                self.logger.info(
+                    f'Reading {len(u_nk)} lines of u_nk from {xvg}')
+                u_nk_list.append(u_nk)
+            except: # pragma: no cover
+                self.logger.warning(
+                    f'Error reading read u_nk from {xvg}.')
+
+            try:
+                dhdl = self._extract_dHdl(xvg, T=self.T)
+                self.logger.info(
+                    f'Reading {len(dhdl)} lines of dhdl from {xvg}')
+                dHdl_list.append(dhdl)
+            except: # pragma: no cover
+                self.logger.warning(
+                    f'Error reading read dhdl from {xvg}.')
+
+        # Sort the files according to the state
+        if len(u_nk_list) > 0:
+            self.logger.info('Sort files according to the u_nk.')
+            column_names = u_nk_list[0].columns.values.tolist()
+            index_list = sorted(range(len(self.file_list)),
+                key=lambda x:column_names.index(
+                    u_nk_list[x].reset_index('time').index.values[0]))
+        else:
+            self.logger.info('Sort files according to the dHdl.')
+            column_names = sorted([dHdl.reset_index('time').index.values[0]
+                                   for dHdl in dHdl_list])
+            index_list = sorted(range(len(self.file_list)),
+                key=lambda x:column_names.index(
+                    dHdl_list[x].reset_index('time').index.values[0]))
```

Comment:
When is the `else` block L123 - L128 being executed? I am a bit worried when a substantial code block is not tested.

(Or is it only the logging line that's missing????)

Reply:
@orbeckst I have removed this code block